### PR TITLE
z_object_switch OK

### PR DESCRIPTION
**Pull request by mzxrules**
Before opening this PR, ensure the following:
- `./format.sh` was run to apply standard formatting.
- `make` successfully builds a matching ROM.
- No new compiler warnings were introduced during the build process.
    - Can be verified locally by running `tools/warnings_count/check_new_warnings.sh`
- New variables & functions should follow standard naming conventions.
- Comments and variables have correct spelling.
---
Also has most functions named.

### CODE REVIEW

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c [hunk 1]
```diff
@@ -28,183 +90,957 @@ const ActorInit Obj_Switch_InitVars = {
...
+void ObjSwitch_InitJntSphCollider(ObjSwitch* this, GlobalContext* globalCtx, ColliderJntSphInit* init) {
+    s32 pad;
+
+    Collider_InitJntSph(globalCtx, &this->colliderJntSph);
+    Collider_SetJntSph(globalCtx, &this->colliderJntSph, &this->dyna.actor, init, this->colliderJntSphElements);
+    Matrix_SetStateRotationAndTranslation(this->dyna.actor.world.pos.x,
+                                          this->dyna.actor.world.pos.y +
+                                              (this->dyna.actor.shape.yOffset * this->dyna.actor.scale.y),
+                                          this->dyna.actor.world.pos.z, &this->dyna.actor.shape.rot);
+    Matrix_Scale(this->dyna.actor.scale.x, this->dyna.actor.scale.y, this->dyna.actor.scale.z, 1);
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c:223]:
```suggestion
    Matrix_Scale(this->dyna.actor.scale.x, this->dyna.actor.scale.y, this->dyna.actor.scale.z, MTXMODE_APPLY);
```

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c [hunk 2]
```diff
@@ -28,183 +90,957 @@ const ActorInit Obj_Switch_InitVars = {
...
+    Actor_SpawnAsChild(&globalCtx->actorCtx, &this->dyna.actor, globalCtx, ACTOR_OBJ_ICE_POLY,
+                       this->dyna.actor.world.pos.x, this->dyna.actor.world.pos.y - 25.0f, this->dyna.actor.world.pos.z,
+                       this->dyna.actor.world.rot.x, this->dyna.actor.world.rot.y, this->dyna.actor.world.rot.z,
+                       0xFF32);
+}
+
+void ObjSwitch_SetSwitchFlagState(ObjSwitch* this, GlobalContext* globalCtx, s32 setFlag) {
+    if (setFlag) {
+        s32 flag = OBJ_SWITCH_GET_SWITCH_FLAG(this->dyna.actor.params);
+        Actor_SetSwitchFlag(globalCtx, flag);
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c:261]:
```suggestion
        s32 flag = OBJ_SWITCH_GET_SWITCH_FLAG(this->dyna.actor.params);

        Actor_SetSwitchFlag(globalCtx, flag);
```

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c [hunk 3]
```diff
@@ -28,183 +90,957 @@ const ActorInit Obj_Switch_InitVars = {
...
+    } else {
+        s32 flag = OBJ_SWITCH_GET_SWITCH_FLAG(this->dyna.actor.params);
+        Actor_UnsetSwitchFlag(globalCtx, flag);
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c:264]:
```suggestion
        s32 flag = OBJ_SWITCH_GET_SWITCH_FLAG(this->dyna.actor.params);

        Actor_UnsetSwitchFlag(globalCtx, flag);
```

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c [hunk 4]
```diff
@@ -28,183 +90,957 @@ const ActorInit Obj_Switch_InitVars = {
...
+    if (OBJ_SWITCH_IS_INVISIBLE(this->dyna.actor.params)) {
+        this->dyna.actor.flags |= 0x80;
+    }
+    if (type == OBJSWITCH_TYPE_EYE) {
+        if (sIsSegmentTableInit == false) {
+            s32 i, j;
+            sIsSegmentTableInit = true;
+
+            for (i = 0; i < 2; i++) {
+                for (j = 0; j < 4; j++) {
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c:402]:
```suggestion
            for (i = 0; i < ARRAY_COUNT(sEyeSwitchTextures); i++) {
                for (j = 0; j < ARRAY_COUNT(*sEyeSwitchTextures); j++) {
```

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c [hunk 5]
```diff
@@ -28,183 +90,957 @@ const ActorInit Obj_Switch_InitVars = {
...
+            Vec3f sp2C;
+            Vec3f sp20;
+
+            Matrix_StatePush();
+            Matrix_RotateY(acActor->world.rot.y, 0);
+            Matrix_InsertXRotation_s(acActor->world.rot.x, 1);
+            Matrix_GetStateTranslationAndScaledZ(1.0f, &sp2C);
+            Matrix_RotateY(actor->shape.rot.y, 0);
+            Matrix_InsertXRotation_s(actor->shape.rot.x, 1);
+            Matrix_InsertZRotation_s(actor->shape.rot.z, 1);
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c:645]:
```suggestion
            Matrix_RotateY(acActor->world.rot.y, MTXMODE_NEW);
            Matrix_InsertXRotation_s(acActor->world.rot.x, MTXMODE_APPLY);
            Matrix_GetStateTranslationAndScaledZ(1.0f, &sp2C);
            Matrix_RotateY(actor->shape.rot.y, MTXMODE_NEW);
            Matrix_InsertXRotation_s(actor->shape.rot.x, MTXMODE_APPLY);
            Matrix_InsertZRotation_s(actor->shape.rot.z, MTXMODE_APPLY);
```

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c [hunk 6]
```diff
@@ -28,183 +90,957 @@ const ActorInit Obj_Switch_InitVars = {
...
+            CollisionCheck_SetOC(globalCtx, &globalCtx->colChkCtx, &this->colliderJntSph.base);
+    }
+}
+
+void ObjSwitch_DrawFloorSwitch(ObjSwitch* this, GlobalContext* globalCtx) {
+    s32 pad[2];
+
+    if (OBJ_SWITCH_GET_SUBTYPE(this->dyna.actor.params) == OBJSWITCH_SUBTYPE_ONCE) {
+        Gfx* opa;
+        OPEN_DISPS(globalCtx->state.gfxCtx);
```
Comment (AngheloAlf) [inline src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c:965]:
```suggestion
        Gfx* opa;

        OPEN_DISPS(globalCtx->state.gfxCtx);
```

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c [hunk 7]
```diff
@@ -5,17 +5,79 @@
...
 
+#define COS_OF_5_PI_DIV_8 -0.38268343f
+
+#define OBJ_SWITCH_GET_TYPE(p) ((p)&7)
+#define OBJ_SWITCH_GET_SUBTYPE(p) (p >> 4 & 7)
+#define OBJ_SWITCH_GET_SWITCH_FLAG(p) ((p) >> 8 & 0x7F)
+#define OBJ_SWITCH_IS_FROZEN(p) ((p) >> 7 & 1)
+#define OBJ_SWITCH_UNSET_FROZEN(p) ((p) &= ~(1 << 7))
+#define OBJ_SWITCH_IS_INVISIBLE(p) ((p) >> 3 & 1)
+#define OBJ_SWITCH_GET_COLOR_ID(z) ((z)&1)
```
Comment (hensldm) [inline src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c:22]: Current standard is params accesses to take in the base actor pointer, as well as be paced in `*.h` file.

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c [hunk 8]
```diff
@@ -5,17 +5,79 @@
  */
 
 #include "z_obj_switch.h"
+#include "objects/gameplay_dangeon_keep/gameplay_dangeon_keep.h"
 
 #define FLAGS 0x00000010
 
 #define THIS ((ObjSwitch*)thisx)
 
+#define COS_OF_5_PI_DIV_8 -0.38268343f
```
Comment (hensldm) [inline src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c:14]: What's everyone's thoughts on putting this in `math.h`?

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c [hunk 9]
```diff
@@ -5,17 +5,79 @@
...
+void ObjSwitch_LargeFloorSwitchRiseUp(ObjSwitch* this, GlobalContext* globalCtx);
+
+void ObjSwitch_DrawFloorSwitch(ObjSwitch* this, GlobalContext* globalCtx);
+void ObjSwitch_DrawRustyFloorSwitch(ObjSwitch* this, GlobalContext* globalCtx);
+void ObjSwitch_DrawVisibleEyeSwitch(ObjSwitch* this, GlobalContext* globalCtx);
+void ObjSwitch_DrawInvisibleEyeSwitch(ObjSwitch* this, GlobalContext* globalCtx);
+void ObjSwitch_DrawEyeSwitch(ObjSwitch* this, GlobalContext* globalCtx);
+void ObjSwitch_DrawCrystalSwitch(ObjSwitch* this, GlobalContext* globalCtx);
+
+static void* sEyeSwitchTextures[][4] = {
```
Comment (hensldm) [inline src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c:74]:
```suggestion
static TexturePtr sEyeSwitchTextures[][4] = {
```
We have a typedef specifically for textures.

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c [hunk 10]
```diff
@@ -28,183 +90,957 @@ const ActorInit Obj_Switch_InitVars = {
...
+            player->linearVelocity = 0.0f;
+        }
+    }
+}
+
+void ObjSwitch_Init(Actor* thisx, GlobalContext* globalCtx) {
+    s32 pad;
+    s32 type;
+    u32 isSwitchFlagSet;
+    ObjSwitch* this = (ObjSwitch*)thisx;
```
Comment (hensldm) [inline src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c:330]:
```suggestion
    ObjSwitch* this = THIS;
```

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c [hunk 11]
```diff
@@ -28,183 +90,957 @@ const ActorInit Obj_Switch_InitVars = {
...
+            } else {
+                ObjSwitch_LargeFloorSwitchUpInit(this);
+            }
+        }
+    }
+}
+
+void ObjSwitch_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    s32 pad;
+    ObjSwitch* this = (ObjSwitch*)thisx;
```
Comment (hensldm) [inline src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c:458]:
```suggestion
    ObjSwitch* this = THIS;
```

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c [hunk 12]
```diff
@@ -28,183 +90,957 @@ const ActorInit Obj_Switch_InitVars = {
...
+    this->dyna.actor.scale.y += 0.074250005f;
+    if (this->dyna.actor.scale.y >= 0.2475f) {
+        ObjSwitch_PlayFootSwitchSfx(this);
+        func_8093AE88(this);
+        ObjSwitch_LargeFloorSwitchUpInit(this);
+    }
+}
+
+void ObjSwitch_Update(Actor* thisx, GlobalContext* globalCtx) {
+    ObjSwitch* this = (ObjSwitch*)thisx;
```
Comment (hensldm) [inline src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c:913]:
```suggestion
    ObjSwitch* this = THIS;
```

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c [hunk 13]
```diff
@@ -28,183 +90,957 @@ const ActorInit Obj_Switch_InitVars = {
...
+
+    CLOSE_DISPS(globalCtx->state.gfxCtx);
+}
+
+void ObjSwitch_Draw(Actor* thisx, GlobalContext* globalCtx) {
+    static ObjSwitchDrawFunc drawFunc[] = {
+        ObjSwitch_DrawFloorSwitch,   ObjSwitch_DrawRustyFloorSwitch, ObjSwitch_DrawEyeSwitch,
+        ObjSwitch_DrawCrystalSwitch, ObjSwitch_DrawCrystalSwitch,    ObjSwitch_DrawFloorSwitch,
+    };
+    ObjSwitch* this = (ObjSwitch*)thisx;
```
Comment (hensldm) [inline src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c:1043]:
```suggestion
    ObjSwitch* this = THIS;
```

**Review by hensldm — CHANGES_REQUESTED** [review]
My thoughts.

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c (hunk 8 above)
Comment (mzxrules) [reply src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c:14]: Seems weird to do for a constant specific to this actor

**Review by hensldm — APPROVED** [review]

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c (hunk 8 above)
Comment (AngheloAlf) [reply src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c:14]: If this constant appears again we can consider moving it to a math header

**Review by AngheloAlf — APPROVED** [review]

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.h [hunk 14]
```diff
@@ -5,15 +5,63 @@
...
+    /* 0x0000 */ DynaPolyActor dyna;
     /* 0x015C */ ObjSwitchActionFunc actionFunc;
-    /* 0x0160 */ char unk_160[0xF8];
+    /* 0x0160 */ s16 floorSwitchReleaseTimer;
+    /* 0x0162 */ s16 disableCrystalSwitchTimer;
+    /* 0x0164 */ s8 eyeTexIndex;
+    /* 0x0165 */ s8 unk_165;
+    /* 0x0168 */ s32 crystalAnimTimer;
+    /* 0x016C */ Color_RGB8 color;
+    /* 0x016F */ u8 unk_16F;
```
Comment (Kenix3) [inline src/overlays/actors/ovl_Obj_Switch/z_obj_switch.h:38]:
```suggestion
    /* 0x016F */ u8 collisionFlags;
```

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.h [hunk 15]
```diff
@@ -5,15 +5,63 @@
...
+    /* 0x0170 */ s8 unk_170;
+    /* 0x0171 */ s8 nextSwitchFlagState;
+    /* 0x0172 */ s8 unk_172;
```
Comment (Kenix3) [inline src/overlays/actors/ovl_Obj_Switch/z_obj_switch.h:41]:
```suggestion
    /* 0x0172 */ s8 isPlayingCutscene;
```

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c [hunk 16]
```diff
@@ -28,183 +82,961 @@ const ActorInit Obj_Switch_InitVars = {
...
+    }
+}
+
+void func_8093AF1C(ObjSwitch* this, s32 arg1) {
+    if (DynaPolyActor_IsInRidingMovingState(&this->dyna)) {
+        this->unk_170 = arg1;
+    }
+}
+
+void func_8093AF54(ObjSwitch* this, GlobalContext* globalCtx) {
```
Comment (Kenix3) [inline src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c:291]: This looks like it's rotating during a cutscene.
```suggestion
void ObjSwitch_TurnPlayer(ObjSwitch* this, GlobalContext* globalCtx) {
```

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.h [hunk 17]
```diff
@@ -5,15 +5,63 @@
...
     /* 0x015C */ ObjSwitchActionFunc actionFunc;
-    /* 0x0160 */ char unk_160[0xF8];
+    /* 0x0160 */ s16 floorSwitchReleaseTimer;
+    /* 0x0162 */ s16 disableCrystalSwitchTimer;
+    /* 0x0164 */ s8 eyeTexIndex;
+    /* 0x0165 */ s8 unk_165;
+    /* 0x0168 */ s32 crystalAnimTimer;
+    /* 0x016C */ Color_RGB8 color;
+    /* 0x016F */ u8 unk_16F;
+    /* 0x0170 */ s8 unk_170;
```
Comment (Kenix3) [inline src/overlays/actors/ovl_Obj_Switch/z_obj_switch.h:39]:
```suggestion
    /* 0x0170 */ s8 turnPlayerState;
```

This is a variable that tracks whether or not the player has been rotated after pressing a floor switch

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c [hunk 18]
```diff
@@ -28,183 +82,961 @@ const ActorInit Obj_Switch_InitVars = {
...
+    }
+}
+
+void ObjSwitch_PlayDiamondSwitchSfx(ObjSwitch* this) {
+    if (this->unk_165 <= 0) {
+        Actor_PlaySfxAtPos(&this->dyna.actor, NA_SE_EV_DIAMOND_SWITCH);
+    }
+}
+
+void func_8093AF1C(ObjSwitch* this, s32 arg1) {
```
Comment (Kenix3) [inline src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c:285]:
```suggestion
void ObjSwitch_SetTurnPlayerState(ObjSwitch* this, s32 arg1) {
```

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c [hunk 19]
```diff
@@ -28,183 +82,961 @@ const ActorInit Obj_Switch_InitVars = {
...
+
+        Flags_UnsetSwitch(globalCtx, flag);
+    }
+}
+
+void ObjSwitch_CrystalUpdateTimer(ObjSwitch* this) {
+    this->crystalAnimTimer++;
+}
+
+void func_8093AE88(ObjSwitch* this) {
```
Comment (Kenix3) [inline src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c:266]:
```suggestion
void ObjSwitch_StopCutscene(ObjSwitch* this) {
```

**Review by Kenix3 — CHANGES_REQUESTED** [review]
Good work on this overall. Left some suggestions to name the rest of the functions.

File: src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c (hunk 16 above)
Comment (mzxrules) [reply src/overlays/actors/ovl_Obj_Switch/z_obj_switch.c:291]: Code doesn't update any rotation variables, so I don't know where you got that idea. I ran a sim on it and it seems like it tries to snap the player to the boundaries of a square that's 48x48/96x96 depending on if it's a regular or big floor switch

**Review by Kenix3 — APPROVED** [review]

**Pull request merged**